protob/pb2py

```diff
@@ -0,0 +1,357 @@
+
+    def write_to_file(self, name, out):
+        # Write generated sourcecode to given file
+        logging.debug("Writing file %s.py" % name)
+        with open(os.path.join(self.out_dir, "%s.py" % name), "w") as f:
+            f.write(AUTO_HEADER)
+            for line in out:
+                f.write(line + "\n")
+
+    def write_to_index(self, statement):
+        if self.module_index:
+            self.module_index.write(statement + "\n")
+
+    def write_classes(self, out_dir, module_index=None):
+        self.out_dir = out_dir
+        self.module_index = module_index
+        self.process_messages(self.messages)
+        self.process_enums(self.enums)
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+
+    parser = argparse.ArgumentParser()
+    # fmt: off
+    parser.add_argument('proto', help="Protobuf definition file")
+    parser.add_argument('out_dir', help="Directory for generated source code")
+    parser.add_argument('-p', '--package', default="trezor", help="Name of protobuf package to extract. Use '-p all' to process all packages")
+    parser.add_argument('-P', '--protobuf-module', default="protobuf", help="Name of protobuf module")
+    parser.add_argument('-l', '--module-index', help="Generate list of modules")
+    parser.add_argument('--message-type', default="MessageType", help="Name of enum with message IDs")
+    parser.add_argument("--protobuf-default-include", default="/usr/include", help="Location of protobuf's default .proto files.")
+    # fmt: on
+    args = parser.parse_args()
+
+    if not os.path.exists(args.proto):
+        print("Proto file does not exist: %s" % args.proto)
+        sys.exit(1)
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        logging.debug("Compiling proto definition from %s", args.proto)
+        proto_dir = os.path.dirname(args.proto) or "."
+        descriptor_set_filename = os.path.join(tmpdir, "DESCRIPTOR_SET")
+        protoc_includes = [
+            "-I%s" % dir for dir in [args.protobuf_default_include, proto_dir]
+        ]
+        subprocess.check_call(
+            [
+                "protoc",
+                "--descriptor_set_out=" + descriptor_set_filename,
+                "--include_imports",
+                args.proto,
+            ]
+            + protoc_includes
+        )
+
+        with open(descriptor_set_filename, "rb") as f:
+            descriptor = Descriptor(
+                f,
+                message_type=args.message_type,
+                package=args.package if args.package != "all" else None,
+                protobuf_module=args.protobuf_module,
+            )
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        module_index_name = os.path.join(tmpdir, "__module_index.py")
+        with open(module_index_name, "w") as module_index:
+            descriptor.write_classes(tmpdir, module_index)
+
+        if args.module_index:
+            os.replace(module_index_name, args.module_index)
```

Comment:
Unless I've misread the code, this should fail because `tmpdir` is unlikely to be on the same filesystem as `args.module_index`

Reply:
argh. not as unlikely in practice but you're right, I assumed that `replace` is fs-safe unlike `rename`

shutils it is then

Reply:
Which operating system do you use? On every system I have used, `/tmp` is a `tmpfs` (in-memory filesystem), so this is failing for me.

Reply:
Linux Mint, /tmp seems to be part of root. I assume this would be the same on Ubuntu, because why would the downstream change this particular thing.

Reply:
Arch Linux uses `tmpfs` and I'm certain other distributions do so.